keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,612 @@
+Service are created by the apiserver rather than by
+kube-controller-manager (and are thus independent of whether the
+Endpoints controller is running or not).
+
+For now, we will not change this (beyond that anyone reading the
+object will now see the deprecation warning).
+
+If, in the future, we decide to disable the Endpoints controller by
+default, we can consider whether it makes sense to stop creating the
+apiserver Endpoints as well. It is possible that we could decide that
+there is a stronger "API" guarantee around the existence of the
+`kubernetes` Endpoints object than there is around Endpoints objects
+in general.
+
+### Endpoints Cleanup
+
+We do not want to leave stale `Endpoints` objects around forever if
+the Endpoints controllers are not running. (This is both a waste of
+disk space and a potential source of confusion since the Endpoints
+objects would quickly become out-of-date and incorrect.)
+
+One possibility would be to just document that administrators should
+delete all existing Endpoints themselves if they are going to disable
+the controllers.
+
+Another possibility would be to create an `endpoints-cleanup`
+controller, that could be enabled explicitly, and document that admins
+should (probably) enable that controller if they disable the others.
+
+Or perhaps the EndpointSlice controller could delete Endpoints objects
+that were more than 24 hours out of date with respect to their
+EndpointSlices?
+
+In all cases, we should probably not automatically delete Endpoints
+that don't looke like they were originally created by the Endpoints
+controller. (That is, we should not delete Endpoints unless they
+correspond to a Service with a selector.)
+
+```
+<<[UNRESOLVED endpoints-cleanup ]>>
+
+Decide what to do here. (In the earlier stages we can just recommend
+manual deletion.)
+
+<<[/UNRESOLVED]>>
+```
+
+### Update Remaining Internal Endpoints Users
+
+The aggregated API server and apiserver service proxying code still
+make use of Endpoints. They will need to be updated to use
+EndpointSlices (and, perhaps, to fall back to Endpoints for a while,
+in case any users were manually creating Endpoints and not mirroring
+them?).
+
+### E2E Test Updates
+
+There are a surprising number of e2e tests that still make use of
+Endpoints, mostly because there was never any active effort to port
+old tests away. These will need to be updated. See the [e2e
+tests](#e2e-tests) section for more details.
+
+### Risks and Mitigations
+
+Obviously if a cluster contains components that read Endpoints
+objects, then disabling the Endpoints controllers would break those
+clusters. Given that the `v1.Endpoints` type would still exist in
+these clusters, the failure mode would not be the component would fail
+entirely with errors; it would just think that the Endpoints it was
+looking for didn't exist ("yet"?).
+
+We would need to mitigate this by helping users to figure out if
```

Comment:
We could write an empty Endpoints object with annotation indicating "this page intentionally left blank because..." ?

Reply:
Potentially something for our recommended ValidatingAdmissionPolicies repo, too: I'm thinking that we define one accepted value for that annotation and then either warn or block on changing the annotation value away from the pattern(s) we validate.

Shouldn't block this KEP.

Reply:
You mean write out a blank Endpoints object for every Service? That seems like it ruins most of the benefit of the KEP.

(And writing out only a single "this page intentionally left blank" object wouldn't help components that were only looking for a specific Endpoints.)

Since we're not talking about changing the _default_ behavior of Kubernetes, I feel like we can dump this on the cluster admin. We will provide tools (metrics, events, etc) to help them figure out if anything in their cluster is using Endpoints, and then they can decide whether or not to disable the Endpoints controller. If they disable the Endpoints controller when their users were depending on it, that's their mistake, just like disabling any other feature that users were using.

Reply:
Writing a single object doesn't need to watch pods, just services.  So it's net simpler.  I don't know if it would have any real impact, though?  At least any client who was waiting for Endpoints to appear would be more likely to find it and maybe  fail in a more obvious way?  And it could maybe go away after some time - the transition is what matters most to me.

More thinking aloud than concrete.

Reply:
That would work (a little bit) for cluster-wide controllers watching Endpoints in all namespaces, but I was worried more about apps/scripts waiting for a _specific_ Endpoints object to appear.

Reply:
I meant to create an empty Endpoints object for each service with a selector.  If you were an app waiting for an Endpoints object, you would see one appear, but it would have no IPs.  My theory being that this would be more likely to result in a human's attention than just never appearing.

But I have NO evidence that it would work.